### PR TITLE
Improve tilde statement type errors

### PR DESCRIPTION
**Pull request by WardBrian**
Currently, no matter what goes wrong when typechecking a statement like `x ~ foo(3)`, the same error is output if `foo` is invalid:
`Ill-typed arguments to '~' statement. No distribution 'foo' was found with the correct signature.`

This is much weaker than for other function errors, which will output valid signatures or ambiguous match results. This PR simply tries to make them look more like function signature errors. It's not perfect (a 100% ideal error might try to re-construct the `~` notation rather than formatting them like the function calls they are under the hood) but it's certainly much more informative than before.

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [ ] If a user-facing facing change was made, the documentation PR is here: <LINK>
    - [x] OR, no user-facing changes were made

## Release notes

Improve type errors generated in `~` statements.

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: src/frontend/Typechecker.ml [hunk 1]
```diff
@@ -1079,54 +1079,49 @@ let verify_valid_sampling_pos loc cf =
...
-    | UniqueMatch (ReturnType UReal, _, _) -> true
-    | _ -> false in
-  let is_sampling_dist_in_math =
-    List.exists ~f:is_name_w_suffix_sampling_dist
-      (Utils.distribution_suffices @ Utils.unnormalized_suffices)
+  let name_w_suffix_sampling_dist suffix =
+    SignatureMismatch.matching_function tenv (name ^ suffix) argumenttypes in
+  let sampling_dists =
+    List.map ~f:name_w_suffix_sampling_dist
+      (Utils.distribution_suffices @ Utils.unnormalized_suffices) in
```
Comment (nhuurre) [inline src/frontend/Typechecker.ml:1086]: As far as I know `matching_function` standardizes the suffix before name lookup so `distribution_suffices` already covers all functions here and `unnormalized_suffices` just adds duplicates to the list.

**Review by nhuurre — CHANGES_REQUESTED** [review]
Thanks, this is a much-needed feature. Are you going to handle bad CDFs in truncation too?

File: src/frontend/Typechecker.ml (hunk 1 above)
Comment (WardBrian) [reply src/frontend/Typechecker.ml]: Huh, you're right. I was just following existing behavior here, which means it's been doing extra work for who-knows how long

**Comment by WardBrian** [comment]
Truncation checks on the CDF are a bit trickier due to the fact that the requirements AND over each other rather than OR. I think it's a less common use case, so while it's still worth fixing it might be better to do a separate PR

**Review by nhuurre — APPROVED** [review]

**Pull request merged**